x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -612,6 +622,65 @@ protected Expression maybeSimplifyNegatable(Expression e) {
 
     }
 
+    /**
+     * Look for numerical literals on the right hand side of a {@link BinaryComparison}s that are out of range for the left hand side data
+     * type and replace the comparison by a literal TRUE or FALSE. This rule should be placed after {@link LiteralsOnTheRight}.
+     */
+    private static class OutOfRangeBinaryComparison extends OptimizerRules.OptimizerExpressionRule<BinaryComparison> {
+
+        OutOfRangeBinaryComparison() {
+            super(TransformDirection.DOWN);
+        }
+
+        @Override
+        protected Expression rule(BinaryComparison bc) {
+            // TODO: add tests with null literals if they do not already exist
+            if (bc.left().dataType().isNumeric() == false || bc.right().dataType().isNumeric() == false) {
```

Comment:
BinaryComparison accepts only numbers - the fact that it reached the optimizer means the plan has been already validated.
See BinaryComparisonSimplification above

Reply:
Fair enough - I wanted to be defensive here, but assuming that `BinaryComparison` only has numerical arguments makes sense.